File: android/src/main/kotlin/com/vanvixi/file_saver_ffi/models/ConflictResolution.kt
package com.vanvixi.file_saver_ffi.models

enum class ConflictResolution(
    val value: Int,
) {
    /**
     * Automatically rename file with (1), (2), (3) suffix pattern
     * Example: file.txt → file (1).txt → file (2).txt
     */
    AUTO_RENAME(0),

    /**
     * Overwrite existing file
     *
     * Platform behavior:
     * - Android 9 and below: Full overwrite capability (requires WRITE_EXTERNAL_STORAGE)
     * - Android 10+: Only overwrites files owned by this app
     *
     * Platform limitation (Android 10+):
     * Due to Scoped Storage, files from other apps cannot be detected.
     * If a file with the same name exists from another app, MediaStore will
     * automatically rename your file instead (e.g., photo.jpg → photo (1).jpg)
     *
     * iOS behavior:
     * - Photos: Own files are overwritten; other apps' files create duplicates
     * - Documents: Full overwrite (each app has isolated sandbox)
     *
     * WARNING: When successful, existing file will be permanently deleted
     */
    OVERWRITE(1),

    /**
     * Fail operation if file exists
     * Returns error code: FILE_EXISTS
     */
    FAIL(2),

    /**
     * Skip operation silently if file exists
     * Returns success with existing file path
     */
    SKIP(3),
    ;

    companion object {
        /**
         * Converts integer value to ConflictResolution
         *
         * @param value Integer value (0-3)
         * @return Corresponding ConflictResolution, defaults to AUTO_RENAME if invalid
         */
        fun fromInt(value: Int): ConflictResolution = entries.find { it.value == value } ?: AUTO_RENAME
    }
}

File: lib/src/models/conflict_resolution.dart
/// Strategy for handling file name conflicts when saving files.
///
/// Defines how to handle the case when a file with the same name
/// already exists at the destination.
enum ConflictResolution {
  /// Automatically rename the file by appending (1), (2), etc.
  ///
  /// Example:
  /// - photo.jpg exists
  /// - Saves as photo (1).jpg
  /// - If photo (1).jpg exists, saves as photo (2).jpg
  ///
  /// This is the default and recommended strategy for most use cases.
  autoRename,

  /// Overwrite the existing file with the new file.
  ///
  /// **Warning**: The existing file will be permanently deleted.
  /// Use with caution.
  overwrite,

  /// Fail the save operation if a file with the same name exists.
  ///
  /// Returns [SaveFailure] with error code "FILE_EXISTS".
  fail,

  /// Skip the save operation if a file with the same name exists.
  ///
  /// Returns [SaveSuccess] with the path of the existing file.
  /// No actual file writing occurs.
  skip,
}
